### PR TITLE
LCOW: API: Add `platform` to /images/create and /build

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

This PR has the API changes described in https://github.com/moby/moby/issues/34617.

_Update 9/13 based on feedback:_
~~Specifically, it adds an HTTP header "X-Requested-Platform" which is a JSON-encoded
OCI Image-spec `Platform` structure to `/images/create` and `/build`~~
Specifically, it adds a query parameter which is a simple string such as `platform=linux/amd64` to `/images/create` and `/build`.

In addition, it renames (almost all) uses of the string variable `platform` (and associated
methods/functions) to `os`, or `operatingSystem` where it clashes with the `os` import. This makes it much clearer to disambiguate with the swarm "platform" which is really `os/arch`. This is a stepping stone to getting the daemon towards fully multi-platform/arch-aware, and makes it clear when "operating system" is being referred to rather than "platform" which is misleadingly used - sometimes in the swarm
meaning, but more often as just the operating system.

### CODE REVIEW

**Comment by jhowardmsft** [comment]
This is what the behaviour with this change (and a CLI updated to support the `--platform` flag) would look like:

This is where the image does not exist locally, and no platform flag has been passed. The engine will default to the host OS which is Windows, but no Windows busybox exists in the registry.
```
PS E:\> docker run --rm busybox echo hello
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
e:\go\src\github.com\docker\cli\binary\docker.exe: cannot download image with operating system "linux" when requesting "windows".
See 'e:\go\src\github.com\docker\cli\binary\docker.exe run --help'.
PS E:\>
```


Now doing the same adding the CLI/API flag. This succeeds.
```
PS E:\> docker run --rm --platform=linux busybox echo hello
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
add3ddb21ede: Pull complete
Digest: sha256:b82b5740006c1ab823596d2c07f081084ecdb32fd258072707b99f52a3cb8692
Status: Downloaded newer image for busybox:latest
hello
```

Now the image exists locally, there's no need to pass the API/CLI flag to run an image based off it:

```
PS E:\> docker run --rm busybox echo hello
hello
```


Delete the image for the next step

```
PS E:\> docker rmi busybox
Untagged: busybox:latest
Untagged: busybox@sha256:b82b5740006c1ab823596d2c07f081084ecdb32fd258072707b99f52a3cb8692
Deleted: sha256:d20ae45477cbc89863fff11d01cdccf28e4ff06ce2eb2f0206ef971b14eaf6c0
Deleted: sha256:6a749002dd6a65988a6696ca4d0c4cbe87145df74e3bf6feae4025ab28f420f2
```

Attempt to pull without an API/CLI flag. This also fails as there is no busybox image for Windows in the registry
```
PS E:\> docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
cannot download image with operating system "linux" when requesting "windows"
```

So adding the flag

```
PS E:\> docker pull --platform=linux busybox
Using default tag: latest
latest: Pulling from library/busybox
add3ddb21ede: Pull complete
Digest: sha256:b82b5740006c1ab823596d2c07f081084ecdb32fd258072707b99f52a3cb8692
Status: Downloaded newer image for busybox:latest
```

**Comment by jhowardmsft** [comment]
And showing the same build step with the API as in  --no-cache -f test3 --platform=linux .


```PowerShell
PS E:\docker\build\lcow> docker build --no-cache -f test3 --platform=linux .
Sending build context to Docker daemon  19.93MB
Step 1/3 : FROM busybox
latest: Pulling from library/busybox
add3ddb21ede: Pull complete
Digest: sha256:b82b5740006c1ab823596d2c07f081084ecdb32fd258072707b99f52a3cb8692
Status: Downloaded newer image for busybox:latest
 ---> d20ae45477cb
Step 2/3 : RUN mkdir /foo
 ---> Running in b83d7de0c946
 ---> 4755c8944ab4
Removing intermediate container b83d7de0c946
Step 3/3 : RUN ls -l /
 ---> Running in 5458cc315cd3
total 36
drwxr-xr-x    2 root     root         12288 Aug 23 00:00 bin
drwxr-xr-x    5 root     root           340 Aug 27 01:34 dev
drwxr-xr-x    1 root     root            60 Aug 27 01:34 etc
drwxr-xr-x    2 root     root          4096 Aug 27 01:34 foo
drwxr-xr-x    2 nobody   nogroup       4096 Aug 23 00:00 home
dr-xr-xr-x  118 root     root             0 Aug 27 01:34 proc
drwxr-xr-x    2 root     root          4096 Aug 23 00:00 root
dr-xr-xr-x   12 root     root             0 Aug 27 01:34 sys
drwxrwxrwt    2 root     root          4096 Aug 23 00:00 tmp
drwxr-xr-x    3 root     root          4096 Aug 23 00:00 usr
drwxr-xr-x    4 root     root          4096 Aug 23 00:00 var
 ---> eaa4cf7ef226
Removing intermediate container 5458cc315cd3
Successfully built eaa4cf7ef226
PS E:\docker\build\lcow>
```

**Comment by jhowardmsft** [comment]
@tonistiigi @dmcgowan @johnstep @mlaventure @stevvooe @vieux @friism PTAL.

This is now ready for review, and can be tested in conjunction with the matching (somewhat hacked as it's dependent on this PR) CLI changes in https://github.com/docker/cli/pull/474. Manual testing results above.

After this, the next step is re-coalescing the image store so that there is a single one in the engine. With that, the Windows daemon should be able to run in dual-LCOW & WCOW modes again, and the environment variable `LCOW_SUPPORTED` can be removed. I'm working on that next step now.

@dnephin @simonferquel @jstarks @darrentstahlmsft @patricklang @thajeztah @crosbymichael FYI.

**Comment by jhowardmsft** [comment]
@stevvooe Note the 3 functions I've added in `pkg\system\lcow.go` which do the translation/parsing/validation of a platform flag passed from the CLI as `os[/arch[/variant]]`. I've switched the engine to use the OCI image-spec `Platform` structure as far as reasonably possible, without implementing the full multi-platform/architecture awareness the engine needs, which, as discussed, is out of scope of just the LCOW work.

I think I will need similar functionality in the containerd package in https://github.com/containerd/containerd/pull/1403 to support these same three functions. However, we can migrate to those at a later date. I've put some pretty clear comments in that the existing functions I've introduced are temporary, and will migrate in the future.

File: api/server/router/image/image_routes.go
Comment (mlaventure) [inline api/server/router/image/image_routes.go]: nit: Can we avoid this extra indent?

For instance, we could name the function return in which case the following defer should work:
```go
defer func() {
    if err != nil {
        if !output.Flushed() {
            return
        }
        output.Write(streamformatter.FormatError(err))
    } 
    err = nil
}()
```

File: api/types/backend/build.go
Comment (mlaventure) [inline api/types/backend/build.go]: Why the rename?

File: api/types/types.go
Comment (mlaventure) [inline api/types/types.go]: Why the rename?

File: builder/dockerfile/containerbackend.go
Comment (mlaventure) [inline builder/dockerfile/containerbackend.go]: ?

File: builder/dockerfile/dispatchers.go
Comment (mlaventure) [inline builder/dockerfile/dispatchers.go]: nit: how about something like:
```go
	imageImage := &image.Image{}
	imageImage.OS = runtime.GOOS
	if runtime.GOOS == "windows" {
		switch b.options.Platform.OS {
		case "windows":
			return nil, errors.New("Windows does not support FROM scratch")
		case "linux":
			if !system.LCOWSupported() {
				return nil, errors.New("LCOW is not supported on this system")
			}
			imageImage.OS = "linux"
		default:
			return nil, errors.Errorf("OS %s is not supported", b.options.Platform.OS)
		}
	}
	return builder.Image(imageImage), nil
```

File: builder/dockerfile/imagecontext.go
Comment (mlaventure) [inline builder/dockerfile/imagecontext.go]: Shouldn't we just pass the platform here?

**Review by mlaventure — CHANGES_REQUESTED** [review]

File: api/types/types.go
Comment (jhowardmsft) [reply api/types/types.go]: @mlaventure It's due to this that I put at the top:

In addition, it renames (almost all) uses of the string variable platform (and associated
methods/functions) to os, or operatingSystem where it clashes with the os import. This makes it much clearer to disambiguate with the swarm "platform" which is really os/arch. This is a stepping stone to getting the daemon towards fully multi-platform/arch-aware, and makes it clear when "operating system" is being referred to rather than "platform" which is misleadingly used - sometimes in the swarm
meaning, but more often as just the operating system.

File: builder/dockerfile/containerbackend.go
Comment (jhowardmsft) [reply builder/dockerfile/containerbackend.go]: Oops. Yes, will fix

File: api/types/backend/build.go
Comment (jhowardmsft) [reply api/types/backend/build.go]: Same response :)

File: api/server/router/image/image_routes.go
Comment (jhowardmsft) [reply api/server/router/image/image_routes.go]: I'm not sure we can that simply. The code can't go down the pull/import path if `GetRequestedPlatform` fails. Unless I'm missing something obvious, which is very possible.

File: builder/dockerfile/imagecontext.go
Comment (jhowardmsft) [reply builder/dockerfile/imagecontext.go]: Actually, the same comment as before. When things are fully multi-platform/arch aware throughout the daemon, I agree. However, for now, the only bit which is actually used is the operating system, hence I'm making that obvious. (Clearly as mentioned before, full multi-platform/arch is out of scope of LCOW directly)

File: builder/dockerfile/dispatchers.go
Comment (jhowardmsft) [reply builder/dockerfile/dispatchers.go]: Sure, will update

**Comment by jhowardmsft** [comment]
@mlaventure comment addressed and rebased.

File: api/swagger.yaml
Comment (stevvooe) [inline api/swagger.yaml]: Why a header with json in it?

**Comment by stevvooe** [comment]
@jhowardmsft Rather than use a header with json, let's define a parameter using the platform selector syntax in containerd/containerd#1403.

**Comment by jhowardmsft** [comment]
@stevvooe. Am I missing what you mean by "as a parameter"? You mean as in `POST /endpoint?foo=bar&platform=.....`? That already was ruled out from a previous comment.....

Otherwise there's only two ways to send "structured data" from the CLI through the API to the daemon. That being either an HTTP header, or JSON encoded in the body.

I can't use the body for this as it `build` uses the body for the build context. Hence the reason for adding it as a header, avoiding a breaking change.

As for encoding, there's really only a handful of choices here. 

a) Pass an un-parsed string and handle it server side. That simply isn't clean.
b) Parse it (and validate) it client side and pass structured data to the server. In terms of encoding structured data, the real two choices open are JSON or base64. JSON IMO seems better here.

I simply went for b). That seems a far better option.

As for using https://github.com/containerd/containerd/pull/1403, as I have commented in the code, the parsing functions are to be updated once that is merged. It isn't yet. However, that PR still bases the underlying platform structure on the OCI Image specs `Platform` definition, which is exactly what I have used here. I don't think it is unreasonable to use that in the API.

File: api/swagger.yaml
Comment (jhowardmsft) [reply api/swagger.yaml]: Answered in larger comment below.

**Comment by dnephin** [comment]
> As for encoding, there's really only a handful of choices here.
> a) Pass an un-parsed string and handle it server side. That simply isn't clean.

This option is the original `os[/arch[/variant]]` right? I actually think that's the best option. The API needs to validate the input anyway, it can't just trust what it receives.

**Comment by stevvooe** [comment]
> You mean as in POST /endpoint?foo=bar&platform=.....? That already was ruled out from a previous comment.....

Which comment?

> As for using containerd/containerd#1403, as I have commented in the code, the parsing functions are to be updated once that is merged. It isn't yet. 

Ok, so let's block this on having that merged. We should use that format throughout the ecosystem to specify platforms. Using the OCI specification is fine, but that ties things to OCI (however, I'd like to see specifiers in OCI).

**Comment by jhowardmsft** [comment]
> Which comment?

@stevvooe - I comes from the discussion in https://github.com/moby/moby/issues/34617#issuecomment-324794410

**Comment by jhowardmsft** [comment]
@stevvooe - Can you comment on the above? I'm trying to determine how you want the API to look in terms of passing the parameter. From conversations offline from @johnstep, it sounds like you do NOT want the JSON.

Which takes me back to what I originally had, as an HTTP API parameter such as `POST /function?platform=os/arch` for example. Can you confirm that?

**Comment by stevvooe** [comment]
@jhowardmsft A query parameter that takes `platform=windows|linux` would be the correct solution and leave us compatible with containerd/containerd#1403.

For json data structures, the OCI platform spec should be used.

**Comment by jhowardmsft** [comment]
@stevvooe To double check before I make the change, JUST `os`? Not `os/arch`?

And by 'query parameter' you mean as in my example`: POST /function?platform=os/arch`

**Comment by jhowardmsft** [comment]
@stevvooe - OK, pushed a second commit which switches the API to a query parameter rather than a JSON header.

**Comment by jhowardmsft** [comment]
Rebased and fixed-up after https://github.com/moby/moby/pull/34258 was merged.

**Comment by jhowardmsft** [comment]
Have opened https://github.com/moby/moby/pull/34858 as a follow-on to this to move to the platform specifier in containerd. Am blocked due to interface changes and the right updates required to moby/moby to complete that. However, these changes can be done as a follow-on to this PR.

In other words, I believe this is now ready for review.

**Comment by jhowardmsft** [comment]
Can we move forward on reviewing this? I have big changes which depend on this (store recoalescing) to allow the daemon to run in dual LCOW/WCOW mode. I'll rebase this tomorrow after the remote file system PR from Akash was merged.

**Comment by jhowardmsft** [comment]
Rebased.

Note the janky error is unrelated - is failing for all PRs currently.

```
02:11:17.935 
02:11:17.935  generated xml file: /go/src/github.com/docker/docker/bundles/test-docker-py/results.xml 
02:11:17.936 =========================== short test summary info ============================
02:11:17.936 XFAIL ../../../../../docker-py/tests/integration/api_container_test.py::CreateContainerTest::test_create_with_init_path
02:11:17.936   init-path removed in 17.05.0
02:11:17.937 XFAIL ../../../../../docker-py/tests/integration/api_container_test.py::CreateContainerTest::test_create_with_storage_opt
02:11:17.937   Not supported on most drivers
02:11:17.937 XFAIL ../../../../../docker-py/tests/integration/models_images_test.py::ImageCollectionTest::test_build_with_error
02:11:17.937   Engine 1.13 responds with status 500
02:11:17.937 SKIP [1] tests/integration/api_image_test.py:270: Doesn't work inside a container - FIXME
02:11:17.937 SKIP [1] tests/integration/api_service_test.py:287: API version is too low (< 1.30)
02:11:17.937 SKIP [1] tests/integration/api_network_test.py:454: API version is too low (< 1.29)
02:11:17.937 SKIP [1] /docker-py/tests/helpers.py:66: Feature requires Docker Engine experimental mode
02:11:17.937 SKIP [1] tests/integration/api_healthcheck_test.py:51: API version is too low (< 1.29)
02:11:17.938 SKIP [1] tests/integration/api_build_test.py:191: API version is too low (< 1.29)
02:11:17.938 SKIP [1] tests/integration/api_service_test.py:300: API version is too low (< 1.27)
02:11:17.938 SKIP [1] /docker-py/tests/integration/api_swarm_test.py:24: Test stalls the engine on 1.12.0
02:11:17.938 SKIP [1] tests/integration/models_services_test.py:84: Makes Swarm unstable?
02:11:17.938 ============== 244 passed, 9 skipped, 3 xfailed in 316.57 seconds ==============
02:11:17.964 ---> Making bundle: .integration-daemon-stop (in bundles/test-docker-py)
02:11:29.165 Clearing AppArmor profiles cache:.
02:11:29.165 All profile caches have been cleared, but no profiles have been unloaded.
02:11:29.166 Unloading profiles will leave already running processes permanently
02:11:29.166 unconfined, which can lead to unexpected situations.

```

**Comment by stevvooe** [comment]
@jhowardmsft I think you should _only_ support os, as I said above, since vendoring the new containerd is going to create other problems.

**Comment by jhowardmsft** [comment]
@stevvooe - OK, that's a trivial change. I'll make it shortly.

**Comment by jhowardmsft** [comment]
@stevvooe - Done. (See `ValidatePlatform` in pkg\system\lcow.go)

**Comment by jhowardmsft** [comment]
Rebased after some builder refactoring was merged recently

**Comment by jhowardmsft** [comment]
Resolved CI failures. All green.

**Comment by jhowardmsft** [comment]
Rebased on master

File: api/types/types.go
Comment (stevvooe) [reply api/types/types.go]: In this case, shouldn't we use an OCI platform spec?

File: api/server/router/image/image_routes.go
Comment (stevvooe) [inline api/server/router/image/image_routes.go]: What err is this? This doesn't look like it gets set.

File: api/server/router/build/build_routes.go
Comment (stevvooe) [inline api/server/router/build/build_routes.go]: :+1:

File: api/server/router/image/image_routes.go
Comment (jhowardmsft) [reply api/server/router/image/image_routes.go]: Good catch, thanks. It was a copy/paste error on my part into the if block above where the platform validation could fail. In this case, we still need to flush the output at the end of this function, but obviously not do the `PullImage` or `ImportImage` calls into the backend. Fixed, pushing it shortly.

File: api/types/types.go
Comment (jhowardmsft) [reply api/types/types.go]: Ah, yes, indeed. `Platform` was added much earlier in the LCOW work. This PR simply renamed it to `OS`. But in the light of the discussions about the OCI spec and specifiers, I agree, this should be the OCI image-spec `Platform`. Will update.

Comment (jhowardmsft) [reply api/types/types.go]: @stevvooe - That's updated now. For the time being, only the OS field is populated as that the only field in the `Container` object currently. As the daemon is moved to be fully multi-platform/arch (etc) aware, as per the F2F discussions, this field can be extended.

**Comment by jhowardmsft** [comment]
Hey @stevvooe - any other issues you can see in here? Thanks :)

**Comment by stevvooe** [comment]
@jhowardmsft I think you need to remove that err check still.

**Comment by jhowardmsft** [comment]
@stevvooe Ah, I see why you say that now. Came back to it with fresh eyes this morning. The intent was that the output was flushed/written, but the backend wasn't called in the case the platform passed in the API was invalid. I've now fixed that, but it requires keeping the error check in.

**Comment by stevvooe** [comment]
@jhowardmsft What can set that error message? Based on the current code, it is always nil, unless I'm missing something.

**Comment by jhowardmsft** [comment]
@stevvooe It's set in the block immediately above it where it's validating what is passed in via the API:

```
...
		platform = system.ParsePlatform(apiPlatform)
		if err = system.ValidatePlatform(platform); err != nil {
			err = fmt.Errorf("invalid platform: %s", err)
		}
	}

	if err == nil {
		if image != "" {
...
```

**Comment by stevvooe** [comment]
LGTM

My that is awkward code.

File: daemon/create.go
Comment (johnstep) [inline daemon/create.go]: s/mean/means/

File: daemon/create.go
Comment (johnstep) [inline daemon/create.go]: The earlier scratch case checks for `system.LCOWSupported()`. Is that required here, or unnecessary there?

File: distribution/pull_v2.go
Comment (johnstep) [inline distribution/pull_v2.go]: bath? Is that short for bailout path?

File: pkg/system/init_windows.go
Comment (johnstep) [inline pkg/system/init_windows.go]: 16299.15? Is the UBR (15, for example) important, and is there a way to deal with it?

File: pkg/system/lcow.go
Comment (johnstep) [inline pkg/system/lcow.go]: No need to change this, since it is temporary and should be short-lived, but I would have checked for `>= 1` first and then nested `>=2` in that, and nested `== 3` in that.

**Review by johnstep — APPROVED** [review]
LGTM

**Comment by jhowardmsft** [comment]
@johnstep (@dmcgowen) - rebased after https://github.com/moby/moby/pull/35090 was merged. Note, I've left the regression introduced by https://github.com/moby/moby/pull/35090#issuecomment-334884586 in there for you guys to come up with a spot fix (with the further caveat I still have reservations about the approach 35090 takes)

**Pull request merged**
